### PR TITLE
New proxies notification

### PR DESCRIPTION
**Pull request by lynx56**
combine_images-22

### CODE REVIEW

File: novawallet/Common/Storage/EntityToModel/ProxyAccountMapper.swift [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+            status: ProxyAccountModel.Status(rawValue: entity.status!)!
+        )
+    }
+
+    func populate(
+        entity _: CoreDataEntity,
+        from _: DataProviderModel,
+        using _: NSManagedObjectContext
+    ) throws {
+        fatalError("populate(entity:) has not been implemented. The mapper is for fetching only")
```
Comment (lynx56) [inline novawallet/Common/Storage/EntityToModel/ProxyAccountMapper.swift:29]: May be implement it later? Things todo
- Add ChainAccount object to ProxyAccountModel
- Add identifier for CDChainAccount and implement mapper for it
- Add MetaAccountModel object to ChainAccountModel

What do you think?

File: novawallet/Modules/WalletsList/Selection/WalletSelectionPresenter.swift [hunk 2]
```diff
@@ -66,17 +66,25 @@ extension WalletSelectionPresenter: WalletSelectionPresenterProtocol {
             !item.isSelected else {
             return
         }
-
+        shouldShowDelegatesUpdates = false
```
Comment (lynx56) [inline novawallet/Modules/WalletsList/Selection/WalletSelectionPresenter.swift:69]: not sure for the solution

File: novawallet/Common/View/WalletSwitchControl.swift [hunk 3]
```diff
@@ -13,6 +13,12 @@ final class WalletSwitchContentView: UIView {
         return imageView
     }()
 
+    let badgeView: UIView = .create {
+        $0.backgroundColor = R.color.colorIconAccent()!
+        $0.layer.cornerRadius = 5
```
Comment (ERussel) [inline novawallet/Common/View/WalletSwitchControl.swift:18]: I would use RoundedView. It has optimised drawing for corners and shadows

File: novawallet/Common/View/WalletSwitchControl.swift [hunk 4]
```diff
@@ -40,6 +52,36 @@ final class WalletSwitchContentView: UIView {
...
+            let frame = CGRect(
+                x: origin.x - holeWidth,
+                y: origin.y - holeWidth,
+                width: width,
+                height: height
+            )
+
+            iconView.cutHole(ovalIn: frame)
+        } else {
+            iconView.layer.mask = nil
```
Comment (ERussel) [inline novawallet/Common/View/WalletSwitchControl.swift:83]: I would add extension method ```removeHole``` for this

File: novawallet/Modules/AssetList/AssetListInteractor.swift [hunk 5]
```diff
@@ -22,18 +22,22 @@ final class AssetListInteractor: AssetListBaseInteractor {
...
     let walletConnect: WalletConnectDelegateInputProtocol
     let assetListModelObservable: AssetListModelObservable
+    let proxyListLocalSubscriptionFactory: ProxyListLocalSubscriptionFactoryProtocol
 
     private var nftSubscription: StreamableProvider<NftModel>?
     private var nftChainIds: Set<ChainModel.Id>?
 
     private var assetLocksSubscriptions: [AccountId: StreamableProvider<AssetLock>] = [:]
     private var locks: [ChainAssetId: [AssetLock]] = [:]
+    private var proxyListSubscription: StreamableProvider<ProxyAccountModel>?
```
Comment (ERussel) [inline novawallet/Modules/AssetList/AssetListInteractor.swift:32]: I would have WalletNotificationService that wraps inside the logic of resolving proxies for current wallet and easy integration to other screens as well as adding new notifications in future

File: novawallet/Modules/WalletsList/Common/WalletIconView.swift [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+    override func layoutSubviews() {
+        super.layoutSubviews()
+
+        guard !networkIconImageView.isHidden else {
+            return
+        }
+
+        iconViewImageView.layoutIfNeeded()
+
+        let width = networkIconImageView.bounds.width + Constants.holeWidth * 2
```
Comment (ERussel) [inline novawallet/Modules/WalletsList/Common/WalletIconView.swift:36]: I believe we have this calculation I would add another method to UIView where ```cutHole``` resides to handle this

File: novawallet/Modules/WalletsList/Common/WalletView.swift [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+    }
+}
+
+extension WalletView {
+    struct ViewModel: Hashable, Equatable {
+        let icon: IdentifiableImageViewModelProtocol?
+        let networkIcon: IdentifiableImageViewModelProtocol?
+        let name: String
+        let subtitle: String
+        let subtitleDetailsIcon: IdentifiableImageViewModelProtocol?
```
Comment (ERussel) [inline novawallet/Modules/WalletsList/Common/WalletView.swift:49]: It is hard to understand for which wallet types which fields are used. I would make enum instead as we discussed: regular and proxy. Each case will contain only fields it is used

File: novawallet/Modules/WalletsList/Common/WalletsListViewModelFactory.swift [hunk 8]
```diff
@@ -100,24 +100,80 @@ extension WalletsListViewModelFactory: WalletsListViewModelFactoryProtocol {
...
+        wallets: [ManagedMetaAccountModel],
+        chains: [ChainModel.Id: ChainModel],
+        locale: Locale
+    ) -> WalletsListViewModel? {
+        guard let chainAccount = wallet.info.chainAccounts.first(where: { $0.proxy != nil }),
+              let proxy = chainAccount.proxy,
+              let proxyWallet = wallets.first(where: { $0.info.has(
+                  accountId: proxy.accountId,
+                  chainId: chainAccount.chainId
+              ) && $0.info.type != .proxied })
```
Comment (ERussel) [inline novawallet/Modules/WalletsList/Common/WalletsListViewModelFactory.swift:136]: I believe we display proxieds in the section. Why do need this check: != .proxied? As I understand ```wallet``` is already proxied.

File: novawallet/Modules/WalletsList/Selection/WalletSelectionInteractor.swift [hunk 9]
```diff
@@ -42,10 +70,34 @@ extension WalletSelectionInteractor: WalletSelectionInteractorInputProtocol {
...
-
                 self?.presenter?.didCompleteSelection()
             case let .failure(error):
-                self?.presenter?.didReceiveError(error)
+                self?.presenter?.didReceive(saveError: error)
+            }
+        }
+    }
+
+    func updateWalletsStatuses() {
```
Comment (ERussel) [inline novawallet/Modules/WalletsList/Selection/WalletSelectionInteractor.swift:80]: Le't move application of these changes to the sync service. In this case running of these operation will not depend on the interactor lifecycle.

File: novawallet/Modules/WalletsList/Selection/WalletSelectionPresenter.swift (hunk 2 above)
Comment (ERussel) [reply novawallet/Modules/WalletsList/Selection/WalletSelectionPresenter.swift:69]: I believe we should call presenter method when controller receives viewWillDisappear. I then presenter call interactor to clear things

**Review by ERussel — COMMENTED** [review]
Screenshot 2023-12-20 at 15 25 17

Sometimes sync adds/revokes that is not actually changes:

Simulator Screenshot - iPhone 14 Pro Max - 2023-12-20 at 15 57 17

File: novawallet/Modules/WalletsList/Common/WalletsListViewModelFactory.swift (hunk 8 above)
Comment (lynx56) [reply novawallet/Modules/WalletsList/Common/WalletsListViewModelFactory.swift:131]: we are looking for proxy for this proxied-wallet

File: novawallet/Modules/Settings/SettingsViewFactory.swift [hunk 10]
```diff
@@ -21,13 +21,19 @@ struct SettingsViewFactory {
             quantityFormatter: NumberFormatter.quantity.localizableResource()
         )
 
+        let proxyNotificationService = WalletNotificationService(
```
Comment (ERussel) [inline novawallet/Modules/Settings/SettingsViewFactory.swift:24]: I wouldn't create it everytime. Let's just add one to service coordinator and pass around

File: novawallet/Modules/WalletsList/Selection/WalletSelectionViewFactory.swift [hunk 11]
```diff
@@ -40,9 +40,23 @@ struct WalletSelectionViewFactory {
             return nil
         }
 
+        let userDataStorageFacade = UserDataStorageFacade.shared
+        let accountRepositoryFactory = AccountRepositoryFactory(storageFacade: userDataStorageFacade)
+        let metaAccountsRepository = accountRepositoryFactory.createManagedMetaAccountRepository(
+            for: nil,
+            sortDescriptors: []
+        )
+        let proxySyncService = ProxySyncService(
```
Comment (ERussel) [inline novawallet/Modules/WalletsList/Selection/WalletSelectionViewFactory.swift:49]: I would use one from service coordinator to not make them collide.

File: novawallet/Modules/WalletsList/Selection/WalletSelectionPresenter.swift [hunk 12]
```diff
@@ -38,10 +38,17 @@ final class WalletSelectionPresenter: WalletsListPresenter {
...
                 return metaAccount.info.type == .proxied && metaAccount.info.chainAccounts.contains {
                     $0.proxy?.status == .new || $0.proxy?.status == .revoked
                 } ? true : false
+            case let .update(_, old, new):
+                guard old.info.type == .proxied, new.info.type == .proxied else {
+                    return false
+                }
+                let oldProxy = old.info.chainAccounts.first(where: { $0.proxy?.isNotActive == true })
+                let newProxy = new.info.chainAccounts.first(where: { $0.proxy != nil })
+                return oldProxy?.proxy?.status != newProxy?.proxy?.status
```
Comment (ERussel) [inline novawallet/Modules/WalletsList/Selection/WalletSelectionPresenter.swift:51]: Didn't understand this logic. If oldProxy is already active and we select it then we receive and update and oldProxy will be nil. If this case nil != newProxy?.proxy?.status and bottomsheet will be shown.

**Pull request merged**